### PR TITLE
K8SPSMDB-636: fill empty credentials in user secrets

### PR DESCRIPTION
**Pull request by pooknull**
[K8SPSMDB-636](https://jira.percona.com/browse/K8SPSMDB-636) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

https://jira.percona.com/browse/K8SPSMDB-636

### CODE REVIEW

File: pkg/controller/perconaservermongodb/secrets.go [hunk 1]
```diff
@@ -158,3 +134,42 @@ func (r *ReconcilePerconaServerMongoDB) reconcileUsersSecret(ctx context.Context
...
+	if cr.CompareVersion("1.13.0") >= 0 {
+		userMap[envMongoDBDatabaseAdminUser] = string(roleDatabaseAdmin)
+		passKeys = append(passKeys, envMongoDBDatabaseAdminPassword)
+	}
+	for k, defaultValue := range userMap {
+		if _, ok := data[k]; !ok {
+			data[k] = []byte(defaultValue)
+			changes = true
+		}
+	}
```
Comment (egegunes) [inline pkg/controller/perconaservermongodb/secrets.go:163]: I suggest the following for clarity:
```suggestion
	for user, role := range userMap {
		if _, ok := data[user]; !ok {
			data[user] = []byte(role)
			changes = true
		}
	}
```

File: pkg/controller/perconaservermongodb/secrets.go [hunk 2]
```diff
@@ -91,58 +91,34 @@ func (r *ReconcilePerconaServerMongoDB) reconcileUsersSecret(ctx context.Context
 		&secretObj,
 	)
 	if err == nil {
+		shouldUpdate, err := fillSecretData(cr, secretObj.Data)
+		if err != nil {
+			return errors.Wrap(err, "failed to fill secret data")
+		}
 		if cr.CompareVersion("1.2.0") < 0 {
-			shouldUpdate := false
```
Comment (nmarukovich) [inline pkg/controller/perconaservermongodb/secrets.go:95]: should we still support 1.2.0 version?

**Review by egegunes — DISMISSED** [review]
LGTM. I just suggested a nitpick but it's not mandatory at all.

File: pkg/controller/perconaservermongodb/secrets.go [hunk 3]
```diff
@@ -158,3 +134,42 @@ func (r *ReconcilePerconaServerMongoDB) reconcileUsersSecret(ctx context.Context
...
+func fillSecretData(cr *api.PerconaServerMongoDB, data map[string][]byte) (changes bool, err error) {
+	if data == nil {
+		data = make(map[string][]byte)
+	}
+	userMap := map[string]string{
+		envMongoDBBackupUser:         string(roleBackup),
+		envMongoDBClusterAdminUser:   string(roleClusterAdmin),
+		envMongoDBClusterMonitorUser: string(roleClusterMonitor),
+		envMongoDBUserAdminUser:      string(roleUserAdmin),
+	}
```
Comment (nmarukovich) [inline pkg/controller/perconaservermongodb/secrets.go:147]: should we use this one 
`MONGODB_DATABASE_ADMIN_USER: databaseAdmin
  MONGODB_DATABASE_ADMIN_PASSWORD: databaseAdmin123456`

File: pkg/controller/perconaservermongodb/secrets.go [hunk 4]
```diff
@@ -158,3 +134,42 @@ func (r *ReconcilePerconaServerMongoDB) reconcileUsersSecret(ctx context.Context
 
 	return nil
 }
+
+func fillSecretData(cr *api.PerconaServerMongoDB, data map[string][]byte) (changes bool, err error) {
```
Comment (inelpandzic) [inline pkg/controller/perconaservermongodb/secrets.go:138]: I suggest staying away from named return types, they introduce a bunch of janky things. To see those its best to see Dave Cheney tip on them: https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_named_return_values

**Comment by JNKPercona** [comment]
| Test name | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/init-deploy.log) |
| one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/one-pod.log) |
| storage | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/storage.log) |
| cross-site-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/cross-site-sharded.log) |
| upgrade-consistency | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/upgrade-consistency.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/demand-backup.log) |
| self-healing | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/self-healing.log) |
| self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/self-healing-chaos.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/limits.log) |
| monitoring-2-0 | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/monitoring-2-0.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/scaling.log) |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/scheduled-backup.log) |
| security-context | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/security-context.log) |
| operator-self-healing | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/operator-self-healing.log) |
| operator-self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/operator-self-healing-chaos.log) |
| rs-shard-migration | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/rs-shard-migration.log) |
| arbiter | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/arbiter.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/service-per-pod.log) |
| liveness | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/liveness.log) |
| smart-update | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/smart-update.log) |
| version-service | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/version-service.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/users.log) |
| data-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/data-sharded.log) |
| demand-backup-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/demand-backup-sharded.log) |
| non-voting | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/non-voting.log) |
| demand-backup-eks-credentials | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/demand-backup-eks-credentials.log) |
| upgrade | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-985/38836753/upgrade.log) |
| We run 27 out of 30|

commit: https://github.com/percona/percona-server-mongodb-operator/pull/985/commits/38836753d466929f48b6cc401b0894dd09b85f09
image: `perconalab/percona-server-mongodb-operator:PR-985-38836753`

File: pkg/controller/perconaservermongodb/secrets.go (hunk 4 above)
Comment (pooknull) [reply pkg/controller/perconaservermongodb/secrets.go]: Thanks for sharing https://github.com/percona/percona-server-mongodb-operator/pull/985/commits/d40a81a9166469829ee2e1a8d6df05c6a67d145b

File: pkg/controller/perconaservermongodb/secrets.go (hunk 1 above)
Comment (pooknull) [reply pkg/controller/perconaservermongodb/secrets.go]: Thanks for sharing

File: pkg/controller/perconaservermongodb/secrets.go (hunk 3 above)
Comment (pooknull) [reply pkg/controller/perconaservermongodb/secrets.go:147]: We have it here https://github.com/percona/percona-server-mongodb-operator/blob/d40a81a9166469829ee2e1a8d6df05c6a67d145b/pkg/controller/perconaservermongodb/secrets.go#L154-L157

**Review by egegunes — APPROVED** [review]

**Review by nmarukovich — APPROVED** [review]

**Review by inelpandzic — APPROVED** [review]

**Pull request merged**